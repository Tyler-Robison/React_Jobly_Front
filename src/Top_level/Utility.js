import React, { useState, useEffect } from "react";
import RouteList from "./RouteList";
import NavBar from "../Nav/NavBar";
import Message from "./Message";
import JoblyApi from "../API/API";
import jwt from 'jsonwebtoken'
import UserContext from "../context/UserContext";
import useLocalStorage from "../customHooks/useLocalStorage";

const Utility = () => {
    const [companies, setCompanies] = useState(null);
    const [jobs, setJobs] = useState(null);
    const [token, setToken] = useLocalStorage('token')
    const [currentUser, setCurrentUser] = useState(null);
    const [msg, setMsg] = useState(null);

    const displayMsg = message => setMsg(message);
    const clearMsg = () => setMsg(null);

    const logout = () => setToken(null);
    const login = token => setToken(token);

    useEffect(() => {
        const loginLogout = async () => {
            if (token && token.length !== 0) {
                const username = jwt.decode(token).username
                // res contains user detail + job applications
                const res = await JoblyApi.getUserInfo(username, token)
                setCurrentUser(res);
            }
            else {
                setCurrentUser(null)
            }
        }
        loginLogout()
    }, [token])

    const providerObj = { currentUser, token }

    return (
        <div>
            <UserContext.Provider value={providerObj}>
                <NavBar logout={logout} />
                <Message msg={msg} />
                <RouteList
                    companies={companies}
                    setCompanies={setCompanies}
                    jobs={jobs}
                    setJobs={setJobs}
                    login={login}
                    displayMsg={displayMsg}
                    clearMsg={clearMsg}
                    setCurrentUser={setCurrentUser}
                />
            </UserContext.Provider>
        </div>
    )
}

export default Utility;