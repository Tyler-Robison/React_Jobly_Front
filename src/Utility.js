import React, { useState, useEffect } from "react";
import RouteList from "./RouteList";
import NavBar from "./NavBar";
import Message from "./Message";
import JoblyApi from "./API";
// import jwt from 'jsonwebtoken'
import UserContext from "./context/UserContext";
import useLocalStorage from "./customHooks/useLocalStorage";

const Utility = () => {
    const [companies, setCompanies] = useState(null);
    const [jobs, setJobs] = useState(null);
    // const [token, setToken] = useState(null);
    const [token, setToken, clearToken] = useLocalStorage('token')
    const [currentUser, setCurrentUser] = useState(null);
    const [msg, setMsg] = useState(null);

    const parseJwt = (token) => {
        try {
            return JSON.parse(atob(token.split('.')[1]));
        } catch (e) {
            return null;
        }
    };

    console.log('current user', currentUser)
    console.log('token', token)

    const displayMsg = (message) => {
        setMsg(message);
    }

    const clearMsg = () => {
        setMsg(null);
    }

    useEffect(() => {
        const loginLogout = async () => {
            if (token && token.length !== 0) {
                const username = parseJwt(token).username
                // res contains user detail
                const res = await JoblyApi.getUserInfo(username, token)
                console.log('token res', res)
                setCurrentUser(res.user);
            } 
            else {
                setCurrentUser(null)
            }
            // console.log('use effect token', token)
        }
        loginLogout()
    }, [token])

    // res.user ->
    // applications: []
    // email: "tylerobison758758@gmail.com"
    // firstName: "Tyler"
    // isAdmin: false
    // lastName: "Robison"
    // username: "janet"

    const logout = () => {
        setToken(null)
    }

    const login = (token) => {
        setToken(token)
    }



    return (
        <div>
            <UserContext.Provider value={currentUser}>
                <NavBar logout={logout} token={token} />
                <Message msg={msg} />
                <RouteList
                    companies={companies}
                    setCompanies={setCompanies}
                    jobs={jobs}
                    setJobs={setJobs}
                    token={token}
                    login={login}
                    displayMsg={displayMsg}
                    clearMsg={clearMsg}
                />
            </UserContext.Provider>
        </div>


    )
}

export default Utility;