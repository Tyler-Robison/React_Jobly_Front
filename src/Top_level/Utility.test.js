import { render, screen } from '@testing-library/react';
import Utility from './Utility';
import { MemoryRouter } from "react-router-dom";


it("renders without crashing", function () {
  <MemoryRouter>
    <Utility />
  </MemoryRouter>
});

it("matches snapshot", function () {
  const container = render(
      <MemoryRouter>
          <Utility />
      </MemoryRouter>
  );
  expect(container.asFragment()).toMatchSnapshot();
});