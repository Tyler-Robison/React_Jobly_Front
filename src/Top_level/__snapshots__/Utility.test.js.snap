// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <div>
      <nav
        class="navbar navbar-expand-lg navbar-light bg-light"
      >
        <div
          class="container"
        >
          <a
            class="navbar-brand"
            href="jobly"
          >
            Jobly
          </a>
          <a
            class="nav-link"
            href="/signup"
          >
            Signup
          </a>
          <a
            class="nav-link"
            href="/login"
          >
            Login
          </a>
        </div>
      </nav>
    </div>
    <h3 />
    <h1>
      Jobly
    </h1>
    <h3>
      All the jobs in one convenient place
    </h3>
    <p />
  </div>
</DocumentFragment>
`;
